zebra-chain/src/block.rs

```diff
@@ -104,22 +115,34 @@ pub struct BlockHeader {
     nonce: [u8; 32],
 
     /// The Equihash solution.
-    // The solution size when serialized should be in bytes ('always
-    // 1344').  I first tried this as a [u8; 1344] but until const
-    // generics land we'd have to implement all our common traits
-    // manually, like in pzec.
-    solution: Vec<u8>,
+    solution: EquihashSolution,
 }
 
 impl ZcashSerialize for BlockHeader {
-    fn zcash_serialize<W: io::Write>(&self, _writer: W) -> Result<(), SerializationError> {
-        unimplemented!();
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), SerializationError> {
+        self.previous_block_hash.zcash_serialize(&mut writer)?;
+        writer.write_all(&self.merkle_root_hash.0[..])?;
+        writer.write_all(&self.final_sapling_root_hash.0[..])?;
+        writer.write_u32::<LittleEndian>(self.time.timestamp() as u32)?;
+        writer.write_u32::<LittleEndian>(self.bits)?;
+        writer.write_all(&self.nonce[..])?;
+        self.solution.zcash_serialize(&mut writer)?;
+        Ok(())
     }
 }
 
 impl ZcashDeserialize for BlockHeader {
-    fn zcash_deserialize<R: io::Read>(_reader: R) -> Result<Self, SerializationError> {
-        unimplemented!();
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        let thing = BlockHeader {
+            previous_block_hash: BlockHeaderHash::zcash_deserialize(&mut reader)?,
+            merkle_root_hash: MerkleTreeRootHash(reader.read_32_bytes()?),
+            final_sapling_root_hash: SaplingNoteTreeRootHash(reader.read_32_bytes()?),
+            time: Utc.timestamp(reader.read_u32::<LittleEndian>()? as i64, 0),
+            bits: reader.read_u32::<LittleEndian>()?,
+            nonce: reader.read_32_bytes()?,
+            solution: EquihashSolution::zcash_deserialize(reader)?,
+        };
+        Ok(thing)
```

Comment:
Could this be just `Ok(BlockHeader { ... })`? The `?` will still work without the variable binding.

Reply:
I think this was leftover from me trying to debug the `?` vs `unwrap()` below, fixing.